### ISSUE TITLE
How do Requests that originate from getasmokealarm and imported via csv get assigned to an Itinerary?

### ISSUE THREAD

**Issue by mgmccarthy** [HTBox/allReady#1567]
cc @MisterJames @tonysurma @stevejgordon 

When importing requests from the `ImportController` or the `RequestApiController`, Requests are being created w/out a corresponding EventId/Event. This makes sense as these requests are coming to us w/out the context of an EventId/Event.

But the only place to look at requests is on the /Admin/Event/Details page:
![image](https://cloud.githubusercontent.com/assets/2428993/20681441/a1ac4aa8-b571-11e6-8b37-28bc72508527.png)

and the query for these Requests is done within the context of an Event. The query that returns the Requests for that given Event looks like this:
```
var requests = await _context.Requests
    .AsNoTracking()
    .Where(rec => rec.EventId == message.EventId)
    .GroupBy(rec => rec.Status)
    .Select(g => new { Status = g.Key, Count = g.Count() })
    .ToListAsync();
```

Because of this, Requests that are imported into the system from the RequestApi or the csv import will never be able to be assigned to an Itinerary b/c there is no way to actually see the Request on our UI.

Is there another Issue/UC that involves building up a UI for Requests with no EventId/Event for eventual assignment to an event so the Request can start to be useful to the system?

Or, maybe there is an UI out there already that I do not know about which allows us to see these Requests that are not created via our UI?

**Comment by tonysurma** [comment]
Requests need to be assigned an eventid upon import or api reception.  Long time ago we talked about having the api key be region unique or some other factor so that there could be a apikey->eventid mapping as there would be different events for different regions.  I recall this discussion in the August timeframe as part of that conference and a bit before.

I do not know what happened to that idea.

For import, we could adjust the UI to allow choosing of an event (dropdown would be itinerary type events that user has access to - so if a org admin for redcross you would see itinerary type events for the red cross).

For API, we need to fall back to mapping apikey-eventid.  Now that being said for December we can make 'hardcoded' (meaning config, or manual entry in database or something not really hardcoded) to map all incoming requests to a single event to get it working for v1 release as we only support API ingestion from getasmokealarm anyways - and then have a v1.1 issue to genericize that so we can have others calling the API.

So bottom line, event import needs UI to choose the event (or we can do same hard code as API) and API needs 'hardcode' of the event to assign requests.

Great find @mgmccarthy (cc @MisterJames @stevejgordon)

**Comment by mgmccarthy** [comment]
@tonysurma, I did not know that before, but it that makes sense. Thanks for explaining.

Looking at what we have now, each Event links to a Location which contains a zip code. We're receiving a region and a zip code from the incoming requests via the API. We could figure out which Event we could associate to the incoming Request and assign it that way. 

I'm not too sure what we would do for a Request whose zip code matches multiple events? This is probably where the region information on the incoming request could come in handy. @MisterJames and I still need to figure out where/how we'll correlate that region data with zip codes in our system, and what model(s) we'll link the data to (Campaign, Event, etc...)

**Comment by mgmccarthy** [comment]
@tonysurma, after a discussion with @MisterJames about issuing an token per region, we've decided against that approach b/c it doesn't really use the token as it's intended to be used. 

The token is issued to provide authentication/authorization functionality to our endpoint. Figuring out whether or not we can service the request is a business operation (not "IT Ops", like authorization) that we should be able to figure out in code.

Regardless, whether we issue multiple tokens per supported region(s)/zip code(s) or go with single token per provider, **the missing piece is how do we identify what regions/zip codes we "support" so we can decide whether or not we can service a request?**

That is what we need to figure out. I'm working with @cecilia-donnelly over at getasmokealarm to examine the data they use to tie together regions/counties/zip codes and based on my understanding, build something on this end that makes sense.

Right now, it looks like we could simply use the zip code on the incoming Request and associate that Request with an Event that is in the zip code using the Location table. Using the zip code is preferable, b/c our system does not have a concept of counties or regions, and I'd like to keep it that way.

If we go with that approach, then the question becomes is there are multiple Events that match the incoming Request's zip code, which Event does the Request get associated with?

If I'm still missing the boat on something, I suggest we climb onto Skype maybe with @OhMcGoo in attendance and talk through this problem so I have a full understanding.

**Comment by EmilyLuijbregts** [comment]
@mgmccarthy and @tonysurma  - just following up on this one. Is it clear what needs to happen? I'm just aware of the December Milestone and this being a "P1" issue.

**Comment by tonysurma** [comment]
@stevejgordon @MisterJames correct me if I am wrong but here is what we discussed on Saturday

For V1, we will have a token for a user (done) that is tied to an org (done).  Upon import via API we will reference the Org they are tied to (tbd) and then store the request associated with that org id (tbd).  When is a user is on an itinerary type event they will be able to pick a button to pull a screen to list requests for that org that aren't assigned to an event (tbd) much / like and close to the manual request work/assignment @stevejgordon has done to date.

The CSV will work the same as their will be a logged in user tied to an org (done) doing the import and upon import via CSV the requests will be associated with the users org (tbd) and the same will be the same above.

This allows us to go live for V1 and not break anything that can't be extended for v1.1 and beyond.

@OhMcGoo can further tell us what regions that will use allready and we can use that in the API per the issue 194 (i believe) in teh get a smoke alarm portal repository

**Comment by stevejgordon** [comment]
@tonysurma That's it exactly. Sorry, I meant to add something to the wiki after Saturday but forgot about. @chris-sargood has submitted the initial PR to add the required OrgId column to the request object so we should be good on the API side there. He's working on the next step to be able to display the requests under #1596.

@chris-sargood - Let me know if you need any support on that next step and we can run through it.

**Comment by chris-sargood** [comment]
@stevejgordon Will do. Did a little bit last night - got something displaying. Going to do a little bit more tonight and grab you tomorrow (if you're in) to double check I've gone about it in the right way.

**Comment by mgmccarthy** [comment]
FYI @stevejgordon and @chris-sargood. 

I submitted PR #1611 last night which, when merged, will put an OrgId of 1 on all incoming API Requests. When this code is merged, we'll then have the correct data conditions for an OrgAdmin user of OrgId 1 to be able to view the Requests via the upcoming UI to be built (Issue #1596)

**Comment by stevejgordon** [comment]
@chris-sargood - Sounds good, thanks.

@mgmccarthy I think we said hardcoding would be okay, but is it possible using the API token to get the org id for the user? That way it's correct if we add a second org?

**Comment by mgmccarthy** [comment]
@stevejgordon, yes that is what we want to do, but first, we need Issue #1586 working.

**Comment by mgmccarthy** [comment]
@stevejgordon @tonysurma 

I think this issue can be closed

**Comment by stevejgordon** [comment]
@mgmccarthy - Thanks Mike. Yep I believe we have this functionality in there now.
